File: 354_max_envelopes/max_envelopes.cpp
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-03-04 09:44:24
 * @Last Modified by: Mengsen.Wang
 * @Last Modified time: 2021-03-04 09:55:09
 */

#include <algorithm>
#include <cassert>
#include <set>
#include <vector>

using namespace std;

int max_envelopes(vector<vector<int>> &&envelopes) {
  sort(envelopes.begin(), envelopes.end(),
       [](const vector<int> &a, const vector<int> &b) -> bool {
         if (a[0] == b[0]) return a[1] > b[1];
         return a[0] < b[0];
       });
  set<int> a;
  for (auto e : envelopes) {
    int v = e[1];
    auto it = lower_bound(a.begin(), a.end(), v);
    if (it != a.end()) {
      a.erase(it);
    }
    a.insert(v);
  }
  return a.size();
}

int main() {
  assert(3 ==
         max_envelopes(vector<vector<int>>{{5, 4}, {6, 4}, {6, 7}, {2, 3}}));
  return 0;
}

File: rust/354_max_envelopes/max_envelopes.rs
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-03-04 09:54:28
 * @Last Modified by: Mengsen.Wang
 * @Last Modified time: 2021-03-04 09:55:57
 */

fn max_envelopes(envelopes: Vec<Vec<i32>>) -> i32 {
    if envelopes.len() == 0 {
        return 0;
    }
    let mut envelopes = envelopes;
    envelopes.sort();
    let mut dp = vec![1; envelopes.len()];
    for i in 1..envelopes.len() {
        for j in 0..i {
            if envelopes[i][0] > envelopes[j][0] && envelopes[i][1] > envelopes[j][1] {
                dp[i] = dp[i].max(dp[j] + 1);
            }
        }
    }
    *dp.iter().max().unwrap()
}

fn main() {
    assert_eq!(
        max_envelopes(vec![vec![5, 4], vec![6, 4], vec![6, 7], vec![2, 3]]),
        3
    )
}
